### DISCUSSION TITLE
Oscillating second derivative computed from gratia package

### DISCUSSION THREAD

**Discussion by RichardYuan04** [gavinsimpson/gratia#325] [Q&A]
Hi there, I have a problem when applying gratia package's derivatives function for my analysis.

I am using only one simple smooth term called BIS(computed by integrating task accuracy and response time), which is a scale for behavioural measurement. And my dependent variable is brain activation value from a cortical area. So the gam is simply:

Brain activation value ~ s(BIS)

Please see the graph for the relationship between a specific brain area (R_V1 region) and BIS:
![Rplot](https://github.com/user-attachments/assets/c49061c2-c401-4e77-a43e-976debed3a0c)

You can see the first derivative is basically positive, and it is indeed a constant when calculated by gratia's derivatives(), around 2.647:
![image](https://github.com/user-attachments/assets/ec64b23a-029b-4701-9a1c-3b323a665bed)

But for the second derivative, I notice that the value start to oscillate around 0, which is very weird and hard to understand for me. I really hope someone could explain why this happens. Please see the chart for second derivative:
![image](https://github.com/user-attachments/assets/0b3972ad-31c6-4cfe-b3ec-29260231b768)

My code for calculating the first and second derivative looks something like this:
```
derivatives(model_all, order = 1, data = WM_Place_2bk, term = sprintf('s(%s)', 'BIS'), type = 'central', unconditional = FALSE)
derivatives(model_all, order = 2, data = WM_Place_2bk, term = sprintf('s(%s)', 'BIS'), type = 'central', unconditional = FALSE)
```

where the model_all is a gam model:
```
formula <- as.formula(paste('R_V1_ROI', "~ s(", 'BIS', ", k = 3)"))
model_all <- gam(formula, data = WM_Place_2bk, method = 'REML')
```
Thank you so much for your attention on this post!

**Comment by gavinsimpson** [comment]
In general there would be two suggestions:

1. you should probably use `m = 3` on the smooth if you want to estimate the second derivative of the resulting smooth. If you use the default, `m = 2`, you are penalizing the derivative you are interested in estimating. That can result in weird things happening, even if you get the next point right,
2. you probably need to make the value of the `eps` argument larger than the default. I plan on improving the default value to be one that adapts to the data, but right now you have to choose this yourself.

**But**, the estimated smooth in your example is linear and thus has 0 second derivative everywhere. Te second derivative is the rate of change in the slope of the estimated smooth. As you smooth doesn't change slope anywhere because it is linear, there is no change in slope and hence the second derivative is 0. That the values are not exactly zero is likely just the loss of precision coming from the floating point arithmetic used by computers (and also the tiny value of `eps` used).

**Reply by RichardYuan04** [reply]
Thank you so much for your suggestions!

**Reply by RichardYuan04** [reply]
I still cannot understand what parameter does the 'm' refer to in the derivatives function, or does 'm' refer to something inside the mgcv's gam function? I also check their manual but didn't find any clue for it. It would be more than appreciative if you could point me in the right way!

Thanks a lot!

**Reply by gavinsimpson** [reply]
`m` is an argument to the `s()` function and is generally used to set details of the penalty for an individual smooth. Not all bases allow this level of configuration (`bs = "cr"` is an example where `m` isn't used), but many do.

Ask this is irrelevant if this estimated smooth is linear (using 1 edf) as that smooth would have 0 second derivative and they oscillations you are seeing are just floating point noise.

**Reply by RichardYuan04** [reply]
Received with thanks!
